Description: Fix an IndexError preventing Genshi for uploading attachments in
 Trac for users with non-English language settings.
Origin: http://genshi.edgewall.org/changeset/1243?format=diff&new=1243
Bug: http://genshi.edgewall.org/ticket/566

--- a/genshi/filters/i18n.py
+++ b/genshi/filters/i18n.py
@@ -1048,7 +1048,13 @@
 
         while parts:
             order, string = parts.pop(0)
-            events = self.events[order].pop(0)
+            events = self.events[order]
+            if events:
+                events = events.pop(0)
+            else:
+                # create a dummy empty text event so any remaining
+                # part of the translation can be processed.
+                events = [(TEXT, "", (None, -1, -1))]
             parts_counter[order].pop()
 
             for event in events:
--- a/genshi/filters/tests/i18n.py
+++ b/genshi/filters/tests/i18n.py
@@ -928,6 +928,18 @@
           """</p></html>""",
           tmpl.generate(first="FIRST", second="SECOND").render())
 
+    def test_translate_i18n_msg_ticket_404_regression(self):
+        tmpl = MarkupTemplate("""<html xmlns:py="http://genshi.edgewall.org/"
+            xmlns:i18n="http://genshi.edgewall.org/i18n">
+          <h1 i18n:msg="name">text <a>$name</a></h1>
+        </html>""")
+        gettext = lambda s: u'head [1:%(name)s] tail'
+        translator = Translator(gettext)
+        translator.setup(tmpl)
+        self.assertEqual("""<html>
+          <h1>head <a>NAME</a> tail</h1>
+        </html>""", tmpl.generate(name='NAME').render())
+
 
 class ChooseDirectiveTestCase(unittest.TestCase):
 
